src/System.Private.CoreLib/shared/System/Double.cs

```diff
@@ -90,8 +90,12 @@ public static unsafe bool IsInfinity(double d)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static unsafe bool IsNaN(double d)
         {
-            long bits = BitConverter.DoubleToInt64Bits(d);
-            return (bits & 0x7FFFFFFFFFFFFFFF) > 0x7FF0000000000000;
+            // A NaN will never equal itself so this is an
+            // easy and efficient way to check for NaN.
+
+            #pragma warning disable CS1718
+            return d != d;
+            #pragma warning restore CS1718
```

Comment:
Well, this code is shared so we automatically get it in Mono (and CoreRT?) so for Mono-LLVM-fast-math it will be `return false` always even for NaN input. Also not sure it won't break IEEE754 for non-LLVM mono (checking) and CoreRT.